### ISSUE TITLE
Building bazel on widows machines flaky

### ISSUE THREAD

**Issue by mhlopko** [bazelbuild/continuous-integration#50]
Strange Windows error was observed today on ci.bazel.io:

> ERROR: C:/jenkins/workspace/Bazel/JAVA_VERSION/1.8/PLATFORM_NAME/windows-x86_64/src/tools/android/java/com/google/devtools/build/android/proto/BUILD:10:1: Generating Java proto_library //src/tools/android/java/com/google/devtools/build/android/proto:serialize_format_proto failed: protoc.exe failed: error executing command bazel-out/host/bin/third_party/protobuf/3.0.0/protoc.exe ... (remaining 3 argument(s) skipped): java.io.IOException: CreateProcess(): %1 is not a valid Win32 application.

http://ci.bazel.io/view/Dashboard/job/Bazel/1174/JAVA_VERSION=1.8,PLATFORM_NAME=windows-x86_64/console

Neither mhlopko, nor laszlocsomor were able to reproduce or find out anything useful.
Re-running the job made the ci green.

**Comment by laszlocsomor** [comment]
I did some investigation.

TL;DR: looks like `ar` is flaky on Windows???

First clue: we somehow generated an empty protoc.exe:

```
$ ls -l /c/bazel_ci/temp/_bazel_SYSTEM/\$7enRRqR/execroot/windows-x86_64/bazel-out/host/bin/third_party/protobuf/3.0.0/protoc.exe
-r-xr-xr-x 1 laszlocsomor None 0 Jan 10 14:45 '/c/bazel_ci/temp/_bazel_SYSTEM/$7enRRqR/execroot/windows-x86_64/bazel-out/host/bin/third_party/protobuf/3.0.0/protoc.exe'
```
vs.
```
$ ls -l /c/bazel_ci/temp/_bazel_SYSTEM/ufohOx12/execroot/windows-x86_64/bazel-out/host/bin/third_party/protobuf/3.0.0/protoc.exe
-r-xr-xr-x 1 laszlocsomor None 11584217 Jan  6 18:04 /c/bazel_ci/temp/_bazel_SYSTEM/ufohOx12/execroot/windows-x86_64/bazel-out/host/bin/third_party/protobuf/3.0.0/protoc.exe
```

It seems that there were some messed up actions upstream. For example, `bazel-out/host/bin/third_party/protobuf/3.0.0/libprotobuf_clib.a` is just 8 bytes with the bad build, but 21M with the good one:

```
$ ls -l /c/bazel_ci/temp/_bazel_SYSTEM/\$7enRRqR/execroot/windows-x86_64/bazel-out/host/bin/third_party/protobuf/3.0.0/libprotobuf_clib.a
-r--r--r-- 1 laszlocsomor None 8 Jan 10 11:56 '/c/bazel_ci/temp/_bazel_SYSTEM/$7enRRqR/execroot/windows-x86_64/bazel-out/host/bin/third_party/protobuf/3.0.0/libprotobuf_clib.a'

$ ls -l /c/bazel_ci/temp/_bazel_SYSTEM/ufohOx12/execroot/windows-x86_64/bazel-out/host/bin/third_party/protobuf/3.0.0/libprotobuf_clib.a
-r--r--r-- 1 laszlocsomor None 21960146 Jan  6 18:03 /c/bazel_ci/temp/_bazel_SYSTEM/ufohOx12/execroot/windows-x86_64/bazel-out/host/bin/third_party/protobuf/3.0.0/libprotobuf_clib.a
```

First few bytes (I guess some non-printable ASCII characters weren't printed):
```
$ head -c16 bad/3.0.0/libprotobuf_clib.a
!<arch>

$ head -c16 good/3.0.0/libprotobuf_clib.a
!<arch>
/
```

Since there's no diff util on the Windows CI machine (not even with msys), <shameless_brag>I had to write <honest>(a rather primitive)</honest> one</shameless_brag>.

Turns out all files under `bazel-out/host/bin/third_party/protobuf/3.0.0/**`  are the same except for:
```
  bazel-out/host/bin/third_party/protobuf/3.0.0/libprotobuf_clib.a
  bazel-out/host/bin/third_party/protobuf/3.0.0/protoc.exe
```

The ".a" file is generated by this step (repro'd on my machine):
```
SUBCOMMAND: # //third_party/protobuf/3.0.0:protobuf_clib [action 'Linking third_party/protobuf/3.0.0/libprotobuf_clib.a']
cd C:/tmp2/_bazel_laszlocsomor/Oh5NykPf/execroot/bazel2
  SET PATH=C:/tools/msys64/usr/bin
  C:/tools/msys64/usr/bin/ar @bazel-out/local-fastbuild/bin/third_party/protobuf/3.0.0/libprotobuf_clib.a-2.params
```

If I rerun this command with the bad and good directory's inputs, I get the same (good) .a file :(
And since all the inputs are the same, the logical conclusion seems to be that the `ar` command failed non-deterministically.. Which isn't helping us at all :(

**Comment by laszlocsomor** [comment]
Here's my super advanced `diff` utility for sake of reproducibility:

```
$ cat diff.cc
#include <stdio.h>

int main(int argc, char** argv) {
  if (argc != 3) {
    fprintf(stderr, "Usage: %s <file1> <file2>\n", argv[0]);
    return 2;
  }

  FILE* fd1 = fopen(argv[1], "rb");
  if (!fd1) {
    fprintf(stderr, "[%s %s] file1 cannot be opened\n", argv[1], argv[2]);
    return 2;
  }

  FILE* fd2 = fopen(argv[2], "rb");
  if (!fd2) {
    fprintf(stderr, "[%s %s] file2 cannot be opened\n", argv[1], argv[2]);
    return 2;
  }

  char buf1[0x400];
  char buf2[0x400];
  bool keep_going = true;
  while (keep_going) {
    size_t size1 = fread(buf1, 0x400, 1, fd1);
    size_t size2 = fread(buf2, 0x400, 1, fd2);
    if (size1 == size2) {
      if (size1 == 0) {
        printf("[%s %s] same\n", argv[1], argv[2]);
        break;
      } else {
        for (int i = 0; i < size1; ++i) {
          if (buf1[i] != buf2[i]) {
            printf("[%s %s] different\n", argv[1], argv[2]);
            keep_going = false;
            break;
          }
        }
      }
    } else {
      printf("[%s %s] different\n", argv[1], argv[2]);
      break;
    }
  }

  fclose(fd1);
  fclose(fd2);

  return 0;
}
```

**Comment by laszlocsomor** [comment]
We cannot repro this, so @mhlopko reimaged the Windows machine. If it happens again, let's reopen and investigate.
